src/foam/nanos/ruler/RuleEngine.java

```diff
@@ -78,15 +75,24 @@ public Object getResult(long ruleId) {
   }
 
   private void applyRules(List<Rule> rules, FObject obj, FObject oldObj) {
+    List<Rule> completedRules = new ArrayList<>();
     for (Rule rule : rules) {
       if ( stops_.get() ) return;
 
       currentRule_ = rule;
       if ( rule.f(getX(), obj, oldObj)
         && rule.getAction() != null
       ) {
-        rule.apply(getX(), obj, oldObj, this);
-        saveHistory(rule, obj);
+        try {
+          rule.apply(getX(), obj, oldObj, this);
+          completedRules.add(rule);
```

Comment:
You could generate less garbage by only allocating completedRules here on the first match.

Reply:
But this is inside for loop. Maybe I misunderstood the comment.

Reply:
Leave completedRules to be null on line 78. Then check if it is null then set it to new ArrayList() here.
This avoids create a new ArrayList even when no rules match.